### PR TITLE
Bug 1862262 - Translations Preferences

### PR DESCRIPTION
**Pull request by ohall-m**
This bug adds a few options for translations preferences on the engine, including:
* (Runtime Setting) Option to get/set translations offer popup
* (Runtime Request) Option to get/set translation language preference
* (Session Request) Option to get/set a site's translation preference

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1862262

### CODE REVIEW

**Comment by ohall-m** [comment]
(Still finishing `(Session Request) Option to get/set a site's translation preference`, but wanted to get a CI test in first.)

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt [hunk 1]
```diff
@@ -837,6 +838,107 @@ class GeckoEngine(
...
     }
 
+    /**
+     * See [Engine.getTranslationsOfferPopup].
+     */
+    override fun getTranslationsOfferPopup(
+        onSuccess: (Boolean) -> Unit,
+        onError: (Throwable) -> Unit,
+    ) {
+        onSuccess(runtime.settings.translationsOfferPopup)
```
Comment (Amejia481) [inline android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:848]: If `translationsOfferPopup` is not asycn (returns righ way) why do we need `onSuccess` and `onError`? Maybe we should just change signature of the function to return `Boolean`?

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt [hunk 2]
```diff
@@ -837,6 +838,107 @@ class GeckoEngine(
...
+    }
+
+    /**
+     * See [Engine.setTranslationsOfferPopup].
+     */
+    override fun setTranslationsOfferPopup(
+        offer: Boolean,
+        onError: (Throwable) -> Unit,
+    ) {
+        runtime.settings.translationsOfferPopup = offer
```
Comment (Amejia481) [inline android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:858]: Same as https://github.com/mozilla-mobile/firefox-android/pull/4628/files#r1411081425

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt [hunk 3]
```diff
@@ -837,6 +838,107 @@ class GeckoEngine(
...
+     */
+    override fun getLanguageSettings(
+        onSuccess: (Map<String, LanguageSetting>) -> Unit,
+        onError: (Throwable) -> Unit,
+    ) {
+        TranslationsController.RuntimeTranslation.getLanguageSettings().then(
+            {
+                if (it != null) {
+                    try {
+                        var result = mutableMapOf<String, LanguageSetting>()
```
Comment (Amejia481) [inline android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:922]: nit: `val` :)

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 4]
```diff
@@ -954,6 +954,62 @@ class GeckoEngineSession(
...
+        onException: (Throwable) -> Unit,
+    ) {
+        if (geckoSession.sessionTranslation == null) {
+            onException(sessionTranslationNotAvailable())
+            return
+        }
+
+        geckoSession.sessionTranslation!!.neverTranslateSiteSetting.then({
+                response ->
+            val errorMessage = "Did not receive a site setting response."
```
Comment (Amejia481) [inline android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:971]: Nit: Maybe we should move `errorMessage` to the `if` statement and it's only used there.

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 5]
```diff
@@ -986,6 +986,23 @@ abstract class EngineSession(
      */
     abstract fun requestTranslationRestore()
 
+    /**
+     * Requests the [EngineSession] retrieve the current site's translation preference.
+     */
+    abstract fun getNeverTranslateSiteSetting(
```
Comment (Amejia481) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt:992]: nit: Not sure if the docs or the name of the function is misleading, could you help me to clarify? 

Doc says get the current site's translation preference but the function indicates `getNeverTranslateSiteSetting`. Should the function be `getTranslationSetting`?

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 6]
```diff
@@ -986,6 +986,23 @@ abstract class EngineSession(
...
+    )
+
+    /**
+     * Requests the [EngineSession] to set the current site's translation preference.
+     */
+    abstract fun setNeverTranslateSiteSetting(
+        setting: Boolean,
+        onResult: () -> Unit,
+        onException: (Throwable) -> Unit,
+    )
```
Comment (Amejia481) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt:1004]: Similar to https://github.com/mozilla-mobile/firefox-android/pull/4628/files#r1411126935

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/LanguageSetting.kt [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.concept.engine.translate
+
+/**
+ * The preferences setting a given language may have on the translations engine.
+ *
+ * @param setting The specified language setting.
+ */
+enum class LanguageSetting(val setting: String) {
```
Comment (Amejia481) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/LanguageSetting.kt:12]: Maybe: `languageSessing`?

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt (hunk 1 above)
Comment (ohall-m) [reply android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt]: Thanks, I wasn't sure of the right pattern here, changed to just Boolean!

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt (hunk 2 above)
Comment (ohall-m) [reply android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:852]: Updated, thanks!

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt (hunk 3 above)
Comment (ohall-m) [reply android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt]: Thanks! (I need to watch for this one more 😄 )

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 4 above)
Comment (ohall-m) [reply android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Updated, thanks!

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt (hunk 5 above)
Comment (ohall-m) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt:992]: Thanks, the docs are misleading, I updated to make it clear it is the `never` setting. (It'll eventually correspond to [this in the Figma for more info too](https://www.figma.com/file/GT6oZLFw32LiCwKl5GOMWD/Full-Page-Translation?type=design&node-id=7202-99696&mode=design&t=VQubXcTivqPukHr6-0).)

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt (hunk 6 above)
Comment (ohall-m) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt:1004]: Updated, thanks!

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/LanguageSetting.kt (hunk 7 above)
Comment (ohall-m) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/LanguageSetting.kt]: Updated, thanks!

**Review by Amejia481 — APPROVED** [review]
Thanks for the fast interaction ♻️  ⚡

**Pull request merged**